### PR TITLE
Remove tcms.core.migrations.0001_django_comments__object_pk

### PR DESCRIPTION
**Pull request by atodorov**
this migration is modifying a 3rd party model and was overriding
app_label to django_comments in order to gain access to the model!

However due to how Django records executed migrations it was
recorded as django_comments|0001... instead of core|0001...! Thus
the migration was seen as still pending and could lead to problems
when applying further migrations.

@tkdchen please review. I think we've got the migration right this time.

### CODE REVIEW

File: CHANGELOG.rst [hunk 1]
```diff
@@ -13,6 +13,8 @@ Change Log
 - Add migration for django_comments
 - Upgrade Django to 1.8.14
 - Upgrade to django-tinymce 2.4.0 (Mr. Senko)
+- Refactor tcms.core.migrations.0001_django_comments__object_pk.
+  PR #157 (Mr. Senko)
```
Comment (tkdchen) [inline CHANGELOG.rst]: Part ``#157`` is not in the summary.

When make new release, I'll make a full list of change history. I think it is not necessary to change this file every time along with a patch.

File: tcms/core/migrations/0001_django_comments__object_pk.py [hunk 2]
```diff
@@ -4,16 +4,41 @@
 from django.db import migrations, models
 
 
+class DjangoCommentsAlterField(migrations.AlterField):
+    """
+        Custom migration operation which alters the `django_comments`
```
Comment (tkdchen) [inline tcms/core/migrations/0001_django_comments__object_pk.py]: I'd like to keep docstring in this format throughout Nitrate.

```
"""summary

More description (could be multiple lines)
"""
```

Please update this dostring.

**Comment by tkdchen** [comment]
Generally LGTM except the two comments.

**Comment by tkdchen** [comment]
In what cases the problem would happen?

**Comment by atodorov** [comment]
> In what cases the problem would happen?

1) execute ./manage.py migrate (possibly for the first time)
2) Because of how Django works (described in the first comment) this migration will be considered as still pending, although it was already applied
3) The next time you try to execute ./manage.py migrate Django will try to apply the same migration and will fail. 

Also during local development, when I start Nitrate via ./manage.py runserver Django was showing a warning (in the console) that I still have unapplied migrations.

**Comment by tkdchen** [comment]
After run ``migrate`` for the first time, ``0001_django_comments__object_pk`` is marked as applied. Then, run ``migrate`` again, there is message

```
  Your models have changes that are not yet reflected in a migration, and so won't be applied.
  Run 'manage.py makemigrations' to make new migrations, and then re-run 'manage.py migrate' to apply them.
```

It seems Django does not know that migration, and I got this

```
./manage.py makemigrations --dry-run -v 3
Migrations for 'django_comments':
  0004_auto_20170403_0004.py:
    - Alter field object_pk on comment
Full migrations file '0004_auto_20170403_0004.py':
# -*- coding: utf-8 -*-
from __future__ import unicode_literals

from django.db import migrations, models


class Migration(migrations.Migration):

    dependencies = [
        ('django_comments', '0003_add_submit_date_index'),
    ]

    operations = [
        migrations.AlterField(
            model_name='comment',
            name='object_pk',
            field=models.TextField(verbose_name='object ID'),
        ),
    ]
```

File ``0004_auto_20170403_0004.py`` is put into ``django_comments/migrations/``.

Django treats it as a model change, and change it migrated ``int`` field back to a ``text`` field. Why does this happen?

**Comment by atodorov** [comment]
> File 0004_auto_20170403_0004.py is put into django_comments/migrations/.
> 
> Django treats it as a model change, and change it migrated int field back to a text field. Why does this happen?


@tkdchen this is because we are modifying a model from a 3rd party application inside our migration. That whole operation is one big mess which we aren't supposed to be doing. What's happening (even when PR is applied) is this:

1) core/migrations/0001_django_comments__object_pk is applied and django_comments.object_pk is converted to int
2) With this PR applied the migration is marked as applied. Without it it is marked as non applied, while at the same time the same migration is recorded as applied but for the django_comments app
3) Because the stock django_comments model is now different the next time you try to run migrate it will detect the difference.
4) If you run makemigrations Django will correctly create a new migration file to reflect the current state of the models. However Django will place this new file under django_comments/migrations because this is where the changes occurred.

Using RawSQL like it was before would help IMO (I will test it in a sec) because the underlying model will be different and we're back to 4).

To make things worse the 'ALTER  COLUMN' statement is not supported on SQLite but using non-SQL migrations Django is able to workaround this. Also there is a bug in handling the migration on PostgreSQL which has been fixed in later versions of Django, see this commit from #171:
https://github.com/Nitrate/Nitrate/pull/171/commits/ef493e2199e3f73d88950b5ee476d5fe7c94afd0


To recap we are modifying django_comments' models OUTSIDE of django_comments and this really confuses the Django migration subsystem. The only alternative I can think about is to take our migration file and tell setup.py to install it under django_comments/migrations. This isn't necessarily better and has its own set of problems.

We can also vendor in the entire django_comments app, add the migration there and carry it around in the Nitrate source code but that defeats the purpose of using packages as dependencies.

The other alternative would be to leave object_pk unchanged, maybe add an index and use int() casting where necessary. I've had something like that in one of my older PRs but then we decided to go ahead with the migration instead.

**Comment by atodorov** [comment]
Update: using RunSQL doesn't seem to suffer from this problem, but it does suffer from the unsupported on SQLite problem, see:
https://travis-ci.org/MrSenko/Nitrate/builds/217867272

I will have to investigate more before I'm certain what is happening inside of Django. It looks like we could special case MySQL and Postgres and use RunSQL here but the model wouldn't know the field changed to int (so do we really gain anything) ?? And need to figure out how to deal with SQLite for local development. I will look at this the next week and update this PR.

**Comment by tkdchen** [comment]
So, for this complicated case, I think this migration can be dropped, as this is an improvement for Nitrate, it's good to have, but not must-have and also not a blocker for using Nitrate. The change in this migration could be documented well as a suggestion. For now, I think it is not worth to spend too much time to fix this. However, I'm glad to see a solution to make it possible.

**Comment by atodorov** [comment]
@tkdchen I've converted this into an empty migration to avoid disrupting existing installations. There are tests which cover add/remove comments for test case runs, I have also verified it manually. Everything seems to be working fine. Also rebased to latest.

Can you merge this PR so I can move on to the next ones ?

**Comment by tkdchen** [comment]
Cool. Thank you very much. I'll look at this latest changes soon, today or tomorrow.

**Comment by tkdchen** [comment]
Tested against a fresh db in steps

```
./manage migrate
./manage migrate --list
./manage migrate
make test
```

Each of them looks work well, there is no issue found. Merge this change. Thank you very much.
